src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -9,121 +9,173 @@ namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine
     using System.Threading.Tasks;
 
     using Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Engine;
 
-    using CrossPlatResources = Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Resources.Resources;
+    using CrossPlatResources = Resources.Resources;
 
     /// <summary>
     /// Orchestrates test session operations for the engine communicating with the client.
     /// </summary>
     public class ProxyTestSessionManager : IProxyTestSessionManager
     {
         private readonly object lockObject = new object();
-        private int parallelLevel;
-        private bool skipDefaultAdapters;
+        private readonly object proxyOperationLockObject = new object();
+        private volatile bool proxySetupFailed = false;
+        private StartTestSessionCriteria testSessionCriteria;
+        private int testhostCount;
+        private TestSessionInfo testSessionInfo;
         private Func<ProxyOperationManager> proxyCreator;
-        private Queue<Guid> availableProxyQueue;
-        private IDictionary<Guid, ProxyOperationManagerContainer> proxyMap;
+        private IList<ProxyOperationManagerContainer> proxyContainerList;
+        private IDictionary<string, int> proxyMap;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ProxyTestSessionManager"/> class.
         /// </summary>
         /// 
-        /// <param name="parallelLevel">The parallel level.</param>
+        /// <param name="criteria">The test session criteria.</param>
+        /// <param name="testhostCount">The testhost count.</param>
         /// <param name="proxyCreator">The proxy creator.</param>
-        public ProxyTestSessionManager(int parallelLevel, Func<ProxyOperationManager> proxyCreator)
+        public ProxyTestSessionManager(
+            StartTestSessionCriteria criteria,
+            int testhostCount,
+            Func<ProxyOperationManager> proxyCreator)
         {
-            this.parallelLevel = parallelLevel;
+            this.testSessionCriteria = criteria;
+            this.testhostCount = testhostCount;
             this.proxyCreator = proxyCreator;
 
-            this.availableProxyQueue = new Queue<Guid>();
-            this.proxyMap = new Dictionary<Guid, ProxyOperationManagerContainer>();
+            this.proxyContainerList = new List<ProxyOperationManagerContainer>();
+            this.proxyMap = new Dictionary<string, int>();
         }
 
         /// <inheritdoc/>
-        public void Initialize(bool skipDefaultAdapters)
+        public virtual bool StartSession(ITestSessionEventsHandler eventsHandler)
```

Comment:
Hmm, that's an interesting idea. I haven't considered this because I saw no need to share the proxies between sessions, do you have a scenario in mind in which this would be a desired approach ? I mostly thought of a session as a thing that stays alive as long as TE wants and is bound to the current set of runsettings + sources. Why would you want to run two separate discoveries/runs for the same source on the same testhost, but in parallel ? Keep in mind you cannot run discovery on one testhost and then, while discovery runs, run test execution on the same testhost.

Reply:
That makes sense, I'll add it then to the list of follow ups in the PR description as I did with the LRU cache.

For now though, that's a potentially breaking change with no immediate benefit and I need more time to properly design such a change, so I'd keep this out of the scope of this PR if you agree too.